File: src/utils/trovoSDK.ts
import * as StellarSdk from '@stellar/stellar-sdk';
import { mnemonicToSeedSync } from 'bip39';
import {Buffer} from "buffer";

const createAccount = (): Account => {
    // create a completely new and unique pair of keys.
    const keyPair = StellarSdk.Keypair.random();   

    /// Returns the human readable account ID of this key pair.
    //return {"publicKey": keyPair.accountId, "secretKey": keyPair.secretSeed};
    return {publicKey:keyPair.publicKey(), secretKey:keyPair.secret()};
}

const signHTTP = (toSign: string, secretKey: string): string => {
    try{
        const keypair = StellarSdk.Keypair.fromSecret(secretKey);
        const encoder = new TextEncoder();
        const list = encoder.encode(toSign);
        const buffer = Buffer.from(list);

        const signedData = keypair.sign(buffer);
        const signedBase64Str = btoa(Array.from(signedData).toString());

        return signedBase64Str;
    }catch(error: any){
        console.log('error signing request', error);
        return '';
    }
}

const importAccount = (secretKey: string): string => {
    try{
        const keypair = StellarSdk.Keypair.fromSecret(secretKey);
        return keypair.publicKey();        
    }catch(error: any){
        console.log('error signing request', error);
        return '';
    }
}

const signBase64Txn = (secretKey: string, transactionXDR: string, networkPassphrase: string): string => {
    try{
        const keypair = StellarSdk.Keypair.fromSecret(secretKey);
        console.log('dkslfsd', keypair, secretKey);
        const txn = new StellarSdk.Transaction(transactionXDR, networkPassphrase);
        console.log('txn', txn);
        const bytes = txn.hash();
        console.log('bytes', bytes);
        
        const signedData = keypair.sign(bytes);
        console.log('bytes', bytes);
        const signedBase64Str = Buffer.from(signedData).toString('base64');
console.log('btoa result', signedBase64Str);
        return signedBase64Str;
    }catch(error: any){
        console.log('error signing request', error);
        return '';
    }
}

const parseSecretKey = (secretKey: string): Account => {
    const keypair = StellarSdk.Keypair.fromSecret(secretKey);

    return {publicKey: keypair.publicKey(), secretKey: keypair.secret()};
};

const getCredsFromPassPhrase = (passphrase: string): Account | null => {
    try {    
        const seed = mnemonicToSeedSync(passphrase);
        const keypair = StellarSdk.Keypair.fromRawEd25519Seed(Buffer.alloc(32, seed));

        return {publicKey: keypair.publicKey(), secretKey: keypair.secret()};
    } catch (error: any) {
        console.log(error);
        return null;        
    }
};

export {createAccount, signHTTP, importAccount, signBase64Txn, parseSecretKey, getCredsFromPassPhrase};

File: lib/functions/trovo-sdk.dart
library trovo_app_sdk_flutter;

import 'dart:convert';
import 'dart:typed_data';

import 'package:stellar_flutter_sdk/stellar_flutter_sdk.dart';

class TrovoWalletSDK {
  Account createAccount() {
    // create a completely new and unique pair of keys.
    KeyPair keyPair = KeyPair.random();

    Account account = Account(keyPair.accountId, keyPair.secretSeed);

    return account;
  }

  String signHTTP({toSign, secretKey}) {
    try {
      KeyPair keyPair = KeyPair.fromSecretSeed(secretKey);

      List<int> list = utf8.encode('$toSign');
      Uint8List bytes = Uint8List.fromList(list);

      // sign with the keypair
      var signedData = keyPair.sign(bytes);

      var signedBase64Str = base64.encode(signedData.toList());

      return signedBase64Str;
    } catch (e) {
      return '';
    }
  }

  importAccount(secretKey) {
    try {
      KeyPair keyPair = KeyPair.fromSecretSeed(secretKey);
      return keyPair.accountId;
    } catch (e) {
      return '';
    }
  }

  signBase64Txn(
    String secretKey,
    String transcationXDR,
    String networkPassphrase,
  ) {
    try {
      KeyPair keyPair = KeyPair.fromSecretSeed(secretKey);

      var txn = AbstractTransaction.fromEnvelopeXdrString(transcationXDR);

      Network network = new Network(networkPassphrase);

      var bytes = txn.hash(network);

      // sign with the keypair
      var signedData = keyPair.sign(bytes);

      var signedBase64Str = base64.encode(signedData.toList());

      return signedBase64Str;
    } catch (e) {
      return '';
    }
  }

  Future<String> generateCredentialsFromPassPhrase() async {
    String mnemonic = await Wallet.generate24WordsMnemonic();
    Wallet wallet = await Wallet.from(mnemonic);
    KeyPair keyPair = await wallet.getKeyPair(index: 1);
    Account(keyPair.accountId, keyPair.secretSeed);
    return mnemonic;
  }

  Future<Account> retrieveCredentialsFromPassPhrase(String passPhrase) async {
    Wallet wallet = await Wallet.from(passPhrase);
    KeyPair keyPair = await wallet.getKeyPair(index: 1);
    return Account(keyPair.accountId, keyPair.secretSeed);
  }

  Account parseSecretKey(secretKey) {
    KeyPair keyPair = KeyPair.fromSecretSeed(secretKey);
    return Account(keyPair.accountId, keyPair.secretSeed);
  }
}

class Account {
  late String publicKey;
  late String secretKey;

  Account(String p, String s) {
    this.publicKey = p;
    this.secretKey = s;
  }

  @override
  String toString() {
    return "Public-Key: " + this.publicKey + " Secret-Key: " + this.secretKey;
  }
}
